### PR TITLE
Add copy constructor for error object

### LINKED ISSUE #6632: Add copy constructor for error object
Sometimes we need to store a reference to an error object somewhere in the code so that we can return it to users. A good example is net.box: if the state machine is in the error state, we should return the error that caused the failure to all user requests. The problem is a user that got an error may link it to another error with [`error.set_prev`](https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_error/error_object/#lua-function.error_object.set_prev), which would modify the original error object stored in the net.box internals as well as errors returned to other users. @Gerold103 [suggested](https://github.com/tarantool/tarantool/pull/6598#discussion_r753809646) to add a copy constructor for the error object so that the net.box state machine would return a copy of the original error to a user, thus ensuring that the original error object stays intact. The new constructor should probably be available both in C and in Lua.

Follow-up https://github.com/tarantool/tarantool/pull/6598

### PR DESCRIPTION
**Pull request by KaitmazianMO**
This patch implements new methods in C and Lua modules 
which make a deep copy of the error object.
The copied object does not contain next error fields of the source object.
These methods allow you to return a copy of the original error to a user, 
thus ensuring that the original error object stays intact.
Closes #6632.

Usage example:

```lua
e1 = box.error.new({code = 1, reason = "1"})
e2 = box.error.new({code = 2, reason = "2"})
e3 = box.error.new({code = 3, reason = "3"})
e4 = box.error.new({code = 4, reason = "4"})
e1:set_prev(e2)
e2:set_prev(e3)
e3:set_prev(e4)
-- e1: e1 -> e2 -> e3 -> e4

copy1 = e1:copy_all()
-- copy1: copy(e1) -> copy(e2) -> copy(e3) -> copy(e4)

copy2 = e2:copy_all()
-- copy2: copy(e2) -> copy(e3) -> copy(e4)

copy3 = e3:copy_one()
-- copy3: copy(e3)

```

### CODE REVIEW

**Comment by kyukhin** [comment]
Hello!
Thanks a lot of for the patchset& First of all, could you please fix the linter?
Here's how it run locally (check all new commits on the current branch):
```bash
$ git clone git@github.com:tarantool/checkpatch.git
$ checkpatch/checkpatch.pl -g master..HEAD
```

**Comment by locker** [comment]
> ```shell
> $ git clone git@github.com:tarantool/checkpatch.git
> $ checkpatch/checkpatch.pl -g master..HEAD
> ```

Note that checkpatch is still work-in-progress and may report false-positive errors, especially for C++ code, so please use your judgement - put the code consistency above checkpatch recommendations. I'll look through the checkpatch output and try to fix false-positive errors.

**Comment by KaitmazianMO** [comment]
> CI still red. See [here](https://github.com/tarantool/tarantool/runs/5564812636?check_suite_focus=true#step:3:115) please.

The master branch [code](https://github.com/tarantool/tarantool/blob/c463e56b90bc8ed21be6161a62df759924daf83c/src/lib/core/exception.h#L94) doesn't use spaces around ':', so I used the same style.  
I think the typedef error is a false-positive error, because there are a lot of examples of one-line typedefs.([typedef](https://github.com/tarantool/tarantool/blob/c463e56b90bc8ed21be6161a62df759924daf83c/src/lib/core/diag.h#L55) in the same file)

File: src/box/error.h [hunk 1]
```diff
@@ -203,6 +203,11 @@ class ClientError: public Exception
 		throw this;
 	}
 
+	virtual struct error *dup() const
+	{
+		return new ClientError;
```
Comment (locker) [inline src/box/error.h:208]: Confusing function name. The function doesn't duplicate the error - it just allocates one. So it should either be named accordingly (alloc?) or actually duplicate the error content.

File: src/lib/core/exception.h [hunk 2]
```diff
@@ -73,6 +73,7 @@ class Exception: public error {
 	int get_line() const { return line; }
 	const char *get_errmsg() const { return errmsg; }
 
+	virtual struct error *dup() const = 0;
```
Comment (locker) [inline src/lib/core/exception.h:76]: Please add a short comment to make checkpatch happy.

File: src/lib/core/exception.h [hunk 3]
```diff
@@ -231,12 +248,20 @@ class CryptoError: public Exception {
...
 
 	virtual void raise() { throw this; }
+
+	virtual struct error *dup() const { return new CryptoError; }
 };
 
 class RaftError: public Exception {
 public:
+	RaftError()
+		:Exception(&type_RaftError, NULL, 0)
```
Comment (locker) [inline src/lib/core/exception.h:258]: Even though other functions in this file, do the same, I think that we should actually have a space after ':' like checkpatch recommends. Please fix. Also it could be written in one line:

```C++
RaftError() : Exception(&type_RaftError, NULL, 0) {}
```

File: src/lua/error.lua [hunk 4]
```diff
@@ -168,11 +180,12 @@ local function error_serialize(err)
...
-    ["raise"] = error_raise;
-    ["match"] = error_match; -- Tarantool 1.6 backward compatibility
-    ["__serialize"] = error_serialize;
-    ["set_prev"] = error_set_prev;
+    ["unpack"]       = error_unpack;
+    ["raise"]        = error_raise;
+    ["copy"]         = error_copy;
+    ["match"]        = error_match; -- Tarantool 1.6 backward compatibility
+    ["__serialize"]  = error_serialize;
+    ["set_prev"]     = error_set_prev;
```
Comment (locker) [inline src/lua/error.lua:188]: Please avoid unrelated changes. There's no need to align this table now. Let's leave it as is and follow the style.

```diff
     ["__serialize"] = error_serialize;
+    ["set_prev"] = error_set_prev;
 }
```

File: src/lib/core/diag.h [hunk 5]
```diff
@@ -53,6 +53,7 @@ struct type_info;
 struct error;
 
 typedef void (*error_f)(struct error *e);
+typedef struct error *(*error_dup_f)(const struct error *e);
```
Comment (locker) [inline src/lib/core/diag.h:56]: Even though it's consistent with other function types defined in this file, I still think we should follow the checkpatch recommendation, because it's more common in Tarantool code:

```C
typedef struct error *
(*error_dup_f)(const struct error *e);
```

File: src/lib/core/exception.cc [hunk 6]
```diff
@@ -59,6 +59,13 @@ exception_log(struct error *error)
 	e->log();
 }
 
+static struct error *
+exception_dup(const struct error *error)
+{
+	Exception *e = (Exception *) error;
+	return e->dup();
```
Comment (locker) [inline src/lib/core/exception.cc:66]: Since this function can be called from C code, it must not raise an exception. Recently, we [decided](https://github.com/tarantool/tarantool/issues/3534) to panic on heap allocation failures so you can use `xmalloc` for error allocations. Alternatively, you can catch a memory allocation exception here or return `out_of_memory` error if `malloc` fails, like `BuildAlloc` does.

File: src/lib/core/diag.c [hunk 7]
```diff
@@ -133,6 +135,62 @@ error_create(struct error *e,
...
+static struct error *
+error_copy_recursive(const struct error *src, struct error *effect)
+{
+	if (src == NULL) {
+		return NULL;
+	}
+	struct error *copy = error_copy_without_linking(src);
+	copy->effect = effect;
+	error_ref(copy);
+	copy->cause = error_copy_recursive(src->cause, copy);
```
Comment (locker) [inline src/lib/core/diag.c:184]: We have limited stack space so please don't use recursion.

Also, I think `error_copy_one` would probably be a better name than `error_copy_without_linking`.

File: test/box/error.test.lua [hunk 8]
```diff
@@ -275,3 +275,30 @@ box.schema.func.drop('runtimeerror')
 --
 err = box.error.new('TestType', 'Message arg1: %s. Message arg2: %u', '1', 2)
 err.message
+
+-- gh-6632: add copy constructor for error object
+--
```
Comment (locker) [inline test/box/error.test.lua:280]: We write all new tests in luatest. Please rewrite in luatest. This test doesn't need `box.cfg` so it should probably live in `test/app-luatest`.

File: test/box/error.result [hunk 9]
```diff
@@ -1060,3 +1060,38 @@ copy1:set_prev(e1)
 copy1:set_prev(e1)
  | ---
  | ...
+-- net.box returns a copy of the internal error
+box.cfg{listen = 3301}
```
Comment (locker) [inline test/box/error.result:1064]: Again, the test should be written in luatest. Since this one requires `box.cfg`, it should reside in `test/box-luatest`.

File: changelogs/unreleased/net.box-request-result-contains-a-copy-of-the-internal-error.md [hunk 10]
```diff
@@ -0,0 +1,3 @@
+## feature/net.box
+
+* Result of request contains a copy of the internal error.
```
Comment (locker) [inline changelogs/unreleased/net.box-request-result-contains-a-copy-of-the-internal-error.md:3]: Before commit 954194a1ca5c2cd1826a8a689663c827313dbbef, the error was copied in Lua, and we never officially released a build with that commit (only beta) so this changelog isn't needed. Please drop.

**Review by locker — CHANGES_REQUESTED** [review]

**Comment by KaitmazianMO** [comment]
@locker, thank you for the huge review!
I've fixed all the problems, but checkpatch can't find the `dup_error` function [comment](https://github.com/KaitmazianMO/tarantool/blob/97a3cbe4b7e87bd67ad06c037484131cf6701c46/src/lib/core/exception.h#L82), looks like a false-positive error.
Please check out the fixes.

File: src/lib/core/exception.h [hunk 11]
```diff
@@ -231,14 +292,42 @@ class CryptoError: public Exception {
...
+	try {
+		struct error *e = new ErrorT;
+		memcpy(e, to_dup, sizeof(*e));
+		e->refs = 0;
+		error_payload_copy(&e->payload, &to_dup->payload);
+		e->cause = NULL;
+		e->effect = NULL;
+		return e;
+	} catch (OutOfMemory *ex) {
+		return ex;
```
Comment (alyapunov) [inline src/lib/core/exception.h:327]: You say that this function 'Duplicates an error with the template type'. But in this line you silently return another error. That's a very strange behavior.

File: src/lib/core/exception.h [hunk 12]
```diff
@@ -231,14 +292,42 @@ class CryptoError: public Exception {
...
 	virtual void raise() { throw this; }
+	virtual struct error *dup() const
+	{
+		return dup_error<RaftError>(this);
+	}
 };
 
+template <class ErrorT>
+struct error *
+Exception::dup_error(const struct error *to_dup)
```
Comment (alyapunov) [inline src/lib/core/exception.h:316]: This function seems to be a copy constructor. In C++ there's already a special way to define a copy constructor. Please use C++ way in C++ code.

**Review by alyapunov — CHANGES_REQUESTED** [review]
Thanks for your effort! See my comments below. The most significant is that we should use C++ semantics in C++ code.

File: src/lib/core/exception.h (hunk 12 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.h:316]: This is not actually a copy constructor, because it duplicates only received error and does not copy any cause and effect errors. I'l modify this function to `Exception *dup_one_exception<ErrorT>(const Exception *e)`, to make it less confusing and more similar to the C++ way.

File: src/lib/core/exception.h (hunk 11 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.h:327]: I'l move try catch block to [`exception_dup`](https://github.com/KaitmazianMO/tarantool/blob/efb1102d9f404cf49cc44296f08432b8ff1268c2/src/lib/core/exception.cc#L62-L67) function. In case of memory allocation error this function will return an `OutOfMemory` error becasue it can be called from C code wich doesn't support exceptions.

File: src/lib/core/exception.h [hunk 13]
```diff
@@ -231,14 +292,42 @@ class CryptoError: public Exception {
...
+{
+	try {
+		struct error *e = new ErrorT;
+		memcpy(e, to_dup, sizeof(*e));
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply src/lib/core/exception.h:320]: I'm using `memcpy` with `struct error` which is a C structure. It seems to be an undefined behavior in case of memcpy with `Exception` class because of it's vtabe. I don't understand why in my case this behavior is undefined, could you please explain me?

File: src/lib/core/exception.h (hunk 12 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.h:316]: I think such a copy constructor might confuse someone. It looks a bit messy if you have a copy constructor in a C++ module that only copies the `Exception` object without any relative nodes, but at the same time in the C module you have the `error_copy` function that copies the `Error` object with all its cause nodes. 
I am going to implenet a copy constructor, which makes a deep copy(with cause nodes) or throws an `OutOfMemory` exception if the memroy allocation fails. Copying without cause nodes can lead to a confusing error. If the `dup` C-function returns the same `OutOfMemory` exception twice you will break the error list while linking error nodes. I can`t copy this exception because I need to allocate memory for its copy.
Alternatively, I can use xmalloc for panic if memory allocation fails, but I want to handle errors by using exceptions (as the master branch code) to avoid different ways of memory allocation and error handling.

Comment (KaitmazianMO) [reply src/lib/core/exception.h:316]: Hi, I'm a little confused with exception handling in this function which is used in C code for calling virtual methods of Exception object by using its base error object. What do you think of this default catch block? @alyapunov 
```C++
static struct error *
exception_dup(const struct error *error, bool *failed)
{
	try {
		Exception *e = (Exception *) error;
		return e->dup();
	} catch (Exception *e) {
		*failed = true;
		return e;
	} catch (...) {
		/** Exceptions are not allowed in C. */
		panic("Unexpected exception while copying");
	}
}
```

**Comment by KaitmazianMO** [comment]
I added the `failed` param to the function `exception_dup`, because I need to have a way to recognize copying fail. It is necessary, because I can get the same `out_of_memory` object and link it twice, so it will break the error list.

File: src/lib/core/exception.h (hunk 11 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.h]: Done. Exception can be raised while copying, so I changed `OutOfMemory` to the base class 'Exception'.

File: src/lib/core/exception.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply src/lib/core/exception.h]: Fixed. Now I'm copying field to field.

File: test/unit/error.c [hunk 14]
```diff
@@ -441,11 +441,54 @@ test_payload_move(void)
...
+	is(strcmp(error_payload_get_str(&p2, "key1"), "value1"), 0, "key1");
+	is(strcmp(error_payload_get_str(&p2, "key2"), "value2"), 0, "key2");
+	is(strcmp(error_payload_get_str(&p2, "key3"), "value3"), 0, "key3");
+
+	error_payload_set_str(&p1, "key4", "value4");
+	is(p2.count, 4, "still 4 fields");
+
+	error_payload_clear(&p1, "key1");
+	is(p2.count, 4, "still 4 fields");
+	is(strcmp(error_payload_get_str(&p2, "key1"), "value1"), 0, "key1");
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply test/unit/error.c:479]: Done.

File: test/unit/error.c [hunk 15]
```diff
@@ -441,11 +441,54 @@ test_payload_move(void)
...
+{
+	header();
+	plan(13);
+
+	struct error_payload p1, p2;
+	error_payload_create(&p1);
+	error_payload_copy(&p2, &p1);
+	ok(p2.count == 0 && p2.fields == NULL, "copied empty");
+
+	error_payload_set_str(&p1, "key", "value");
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply test/unit/error.c:455]: Done.

File: src/lib/core/error_payload.c [hunk 16]
```diff
@@ -297,6 +297,34 @@ error_payload_move(struct error_payload *dst, struct error_payload *src)
...
+	if (src->fields) {
+		struct error_field **fields = xmalloc(sizeof(*fields) * count);
+		for (int i = 0; i < count; ++i) {
+			const struct error_field *to_copy = src->fields[i];
+			uint32_t name_size = strlen(to_copy->name) + 1;
+			uint32_t data_offset = offsetof(struct error_field,
+							name[name_size]);
+			uint32_t total = data_offset + to_copy->size;
+
+			/** extra byte is needed to terminate str-type data. */
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply src/lib/core/error_payload.c]: Fixed.

File: changelogs/unreleased/net.box-request-result-contains-a-copy-of-the-internal-error.md (hunk 10 above)
Comment (KaitmazianMO) [reply changelogs/unreleased/net.box-request-result-contains-a-copy-of-the-internal-error.md]: Done.

File: test/box/error.result (hunk 9 above)
Comment (KaitmazianMO) [reply test/box/error.result]: Done. See test/box-luatest/gh_6632_result_returns_a_copy_of_the_internal_error_test.lua.

File: test/box/error.test.lua (hunk 8 above)
Comment (KaitmazianMO) [reply test/box/error.test.lua]: Done. See test/app-luatest/gh_6632_add_copy_constructor_for_error_object_test.lua.

File: src/lib/core/exception.cc (hunk 6 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.cc]: Done. Now `exception_dup` returns a copy of the error or the error that caused fail.

File: src/lib/core/diag.h (hunk 5 above)
Comment (KaitmazianMO) [reply src/lib/core/diag.h]: Done.

File: src/lib/core/exception.h (hunk 3 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.h]: Fixed.

File: src/lib/core/exception.h (hunk 2 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.h]: Done.

File: changelogs/unreleased/gh-6632-add-copy-constructor-for-error-object.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+e2:set_prev(e3)
+e3:set_prev(e4)
+--- e1: e1 -> e2 -> e3 -> e4
+
+copy1 = e1:copy()
+--- copy1: copy(e1) -> copy(e2) -> copy(e3) -> copy(e4)
+
+copy2 = e2:copy()
+--- copy2: copy(e2) -> copy(e3) -> copy(e4)
+```
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply changelogs/unreleased/gh-6632-add-copy-constructor-for-error-object.md]: Done.

File: src/lib/core/error_payload.h [hunk 18]
```diff
@@ -45,6 +45,10 @@ struct error_payload {
 void
 error_payload_create(struct error_payload *p);
 
+/** Create error payload by another error payload. */
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply src/lib/core/error_payload.h]: Fixed.

File: src/lib/core/error_payload.c [hunk 19]
```diff
@@ -297,6 +297,31 @@ error_payload_move(struct error_payload *dst, struct error_payload *src)
...
+
+		for (int i = 0; i < count; ++i) {
+			struct error_field *f = src->fields[i];
+			uint32_t name_size = strlen(f->name) + 1;
+			uint32_t data_offset = offsetof(struct error_field,
+							name[name_size]);
+			uint32_t total = data_offset + f->size;
+
+			dst->fields[i] = xmalloc(total);
+			memcpy(dst->fields[i], f, total);
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply src/lib/core/error_payload.c]: Done.

**Comment by KaitmazianMO** [comment]
@locker, Please check the checkpatch error list. Spaces around "&" cause other errors.

**Comment by KaitmazianMO** [comment]
Hello!
I fixed merge conflicts. There are still checkpatch errors with references in C++ code, but @locker said I can ignore them. Can anyone review the patch?

**Comment by alyapunov** [comment]
Sorry for delay! It was a hard period of time for our team.
I you would like to continue, please contact with @CuriousGeorgiy. We discussed your patch and it seems that we found a way to make it a bit more simple.
Also sorry for our error/Exception implementation, it's a piece of shit and it is really hard to implement new features without tears.

File: src/lib/core/exception.cc [hunk 20]
```diff
@@ -118,6 +136,34 @@ Exception::Exception(const struct type_info *type_arg, const char *file,
 		     exception_log, type_arg, file, line);
 }
 
+Exception *Exception::deep_copy() const
+{
+	Exception *head = NULL;
+	try {
+		head = dup();
+		head->cause = cause;
```
Comment (alyapunov) [inline src/lib/core/exception.cc:144]: You should set `head->cause` only to duplicated cause. This line of code:
1. Meaningless. Head is not intended to hold this pointer.
2. Can confuse the reader of that code.
3. Leads to severe bug. Imagine the next `dup` will throw, you'll catch it and start to cleanup. You will delete `cause` in current Exception object, leaving broken pointer.

File: src/lib/core/exception.cc [hunk 21]
```diff
@@ -118,6 +136,34 @@ Exception::Exception(const struct type_info *type_arg, const char *file,
...
+
+		struct error *err = head;
+		struct error *cause;
```
Comment (alyapunov) [inline src/lib/core/exception.cc:147]: Please use a name different from member name.

**Review by alyapunov — CHANGES_REQUESTED** [review]
Thanks for the patch, see my comments below.
Also please take a look to commit messages, it seems that you have two commits that "Closes #6632", that is logically wrong. The first commit is not related to any issue, I think it should be "Part of #6632".
And "NO_CHANGELOG=feature was not released" - I can't understand the reason. We always write change changelogs before a feature was not released. We do it for the further release to publish all changes of the release.. Perhaps we should add a proper changelog here.

**Comment by alyapunov** [comment]
Also our lint complains about extern. Pehraps you should declare this function in error.h and implement it in exception.cc

File: src/lib/core/exception.cc (hunk 21 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.cc]: Ok. I changed the names.

File: src/lib/core/exception.cc (hunk 20 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.cc]: Thank you for the detailed explanation.
Seems to be fixed.

File: src/lib/core/exception.cc [hunk 22]
```diff
@@ -37,11 +37,18 @@
 #include "fiber.h"
 #include "reflection.h"
 
+/** out_of_memory::size is zero-initialized by the linker. */
+static OutOfMemory
+out_of_memory(__FILE__, __LINE__, sizeof(OutOfMemory), "malloc", "exception");
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply src/lib/core/exception.cc]: Fixed.

File: src/box/error.h [hunk 23]
```diff
@@ -218,6 +218,10 @@ class ClientError: public Exception
 	{
 	}
 
+	ClientError(const ClientError& other) : Exception(other) {}
```
[the remark this answers is not in the document]
Comment (KaitmazianMO) [reply src/box/error.h]: I removed redundant copy constructors.

File: src/box/error.h (hunk 1 above)
Comment (KaitmazianMO) [reply src/box/error.h]: Now there is a `copy()` method that calls an appropriate copy constructor.
Shouldn't it be resolved? If you don't agree please explain the reasons.

File: src/lib/core/exception.h (hunk 11 above)
Comment (KaitmazianMO) [reply src/lib/core/exception.h]: There is no this function anymore.
